gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/SchedulerLeaseAlgoHandler.java

```diff
@@ -0,0 +1,179 @@
+ * previous lease owner's completion status after the lease should expire to ensure the event is handled in failure
+ * cases.
+ */
+public class SchedulerLeaseAlgoHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(SchedulerLeaseAlgoHandler.class);
+  private final int staggerUpperBoundSec;
+  private static Random random = new Random();
+  protected MultiActiveLeaseArbiter multiActiveLeaseArbiter;
+  protected JobScheduler jobScheduler;
+  protected SchedulerService schedulerService;
+  protected DagActionStore dagActionStore;
+  private MetricContext metricContext;
+  private ContextAwareMeter numLeasesCompleted;
+  @Inject
+  public SchedulerLeaseAlgoHandler(Config config, MultiActiveLeaseArbiter leaseDeterminationStore,
+      JobScheduler jobScheduler, SchedulerService schedulerService, DagActionStore dagActionStore) {
+    this.staggerUpperBoundSec = ConfigUtils.getInt(config,
+        ConfigurationKeys.SCHEDULER_STAGGERING_UPPER_BOUND_SEC_KEY,
+        ConfigurationKeys.DEFAULT_SCHEDULER_STAGGERING_UPPER_BOUND_SEC);
+    this.multiActiveLeaseArbiter = leaseDeterminationStore;
+    this.jobScheduler = jobScheduler;
+    this.schedulerService = schedulerService;
+    this.dagActionStore = dagActionStore;
+    this.metricContext = Instrumented.getMetricContext(new org.apache.gobblin.configuration.State(ConfigUtils.configToProperties(config)),
+        this.getClass());
+    this.numLeasesCompleted = metricContext.contextAwareMeter(RuntimeMetrics.GOBBLIN_SCHEDULER_LEASE_ALGO_HANDLER_NUM_LEASES_COMPLETED);
+  }
+
+  /**
+   * This method is used in the multi-active scheduler case for one or more hosts to respond to a flow action event
+   * by attempting a lease for the flow event and processing the result depending on the status of the attempt.
+   * @param jobProps
+   * @param flowAction
+   * @param eventTimeMillis
+   * @throws IOException
+   */
+  public void handleNewSchedulerEvent(Properties jobProps, DagActionStore.DagAction flowAction, long eventTimeMillis)
+      throws IOException {
+    LeaseAttemptStatus leaseAttemptStatus =
+        multiActiveLeaseArbiter.tryAcquireLease(flowAction, eventTimeMillis);
+    // TODO: add a log event or metric for each of these cases
+    switch (leaseAttemptStatus.getClass().getSimpleName()) {
+      case "LeaseObtainedStatus":
+        finalizeLease((LeaseObtainedStatus) leaseAttemptStatus, flowAction);
+        break;
+      case "LeasedToAnotherStatus":
+        scheduleReminderForEvent(jobProps, (LeasedToAnotherStatus) leaseAttemptStatus, flowAction, eventTimeMillis);
+        break;
+      case "NoLongerLeasingStatus":
+        break;
+      default:
+    }
+  }
+
+  // Called after obtaining a lease to persist the flow action to {@link DagActionStore} and mark the lease as done
+  private boolean finalizeLease(LeaseObtainedStatus status, DagActionStore.DagAction flowAction) {
```

Comment:
name this `launchFlow`?

Reply:
this could be a variety of actions including launch, retry, trigger, etc... which is hard to encapsulate with launchFlow. I changed to `persistFlowAction`

Reply:
that name is fine.  "perform action" is too.

since the class was named `SchedulerLeaseAlgoHandler`, I was presuming that the expected/only action would be launching the flow... but perhaps this is more general than I perceived...